File: 240926/기울어진 직사각형/slanted-rectangle.py
def dfs(x, y, score, dir):
    global answer, r, c, m
    if(x < 0 or x >= n or y < 0 or y >= n):
        return
    
    if (x == r and y == c and dir == 4):
        if answer < score:
            answer = score

        return
    

    score += li[x][y]
    end = 4 if dir == 4 else dir + 1
    for d in range(dir, end + 1):
        nx = x + m[d][0]
        ny = y + m[d][1]

        dfs(nx, ny, score, d)


m = {
    1 : (-1 ,1),
    2 : (-1, -1),
    3 : (1, -1),
    4 : (1, 1)
}


n = int(input())
li = [list(map(int, input().split())) for _ in range(n)]
answer = 0

for r in range(2, n):
    for c in range(1, n - 1):
        tmp = 0
        dfs(r, c, tmp, 1)

print(answer)

File: 240926/기울어진 직사각형/slanted-rectangle.java
import java.util.*;

public class Main {
    static int n;
    static int[][] li;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        li = new int[n][n];
        for (int r = 0; r < n; r++){
            for (int c = 0; c < n; c++){
                li[r][c] = sc.nextInt();
            }
        }

        int answer = 0;
        for (int r = 2; r < n; r++){
            for(int c = 1; c < n - 1; c++){
                int tmp = li[r][c] + li[r - 1][c + 1] + li[r - 2][c] + li[r - 1][c - 1];

                int x1 = r - 1; int y1 = c + 1;
                int x2 = r - 2; int y2 = c;

                while(true){
                    x1 -= 1; y1 += 1;
                    x2 -= 1; y2 += 1;

                    if (x1 >= 0 && x1 < n && x2 >= 0 && x2 < n
                    && y1 >= 0 && y1 < n && y2 >= 0 && y2 < n){
                        tmp += (li[x1][y1] + li[x2][y2]);
                    }
                    else{
                        break;
                    }
                }

                if (tmp > answer){
                    answer = tmp;
                }

            }
        }

        System.out.println(answer);

    }
}